### PR TITLE
elm: refactor file structure

### PR DESCRIPTION
**Pull request by PedroHLC**
## Description of changes

- The `elm/default.nix` was an unreadable spaghetti, so I split it into a directory per-package. While grouping them where possible (`lib`, `ghc8_10`, `ghc9_2`, `ghc9_10` and `node`).
- Patches moved to their specific-package's directory.
- The original file even had a broken indentation (changing it in the middle of an attrset), so I reformatted all new `**/default.nix`.
- Synced the update-script to not undo the changes from https://github.com/NixOS/nixpkgs/commit/ad11438d0913939b07ac95e6871d76b3e83ff713
- Deleted `pkgs/development/compilers/elm/packages/bimap.nix`, it wasn't being used anywhere.
- Added more `cabal2nix` to `update.sh`
- Moved manual changes from originally `cabal2nix`-generated files to overrides.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format-test-lib.nix [hunk 1]
```diff
@@ -1,6 +1,20 @@
...
+, hspec-golden
+, lib
+, mtl
+, split
+, tasty
+, tasty-discover
+, tasty-hspec
+, tasty-hunit
+, text
 }:
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format-test-lib.nix:18]: these files are generated by cabal2nix and are not meant to be edited by hand. What did you use to format them? If you did this by hand than I think we should rollback the change as it will be undone next time script we run the code gen.

File: pkgs/development/compilers/elm/packages/node/default.nix [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  elm-verify-examples =
+    let
+      patched = patchBinwrap [ elmi-to-json ] nodePkgs.elm-verify-examples // {
+        meta = with lib; nodePkgs.elm-verify-examples.meta // {
+          description = "Verify examples in your docs";
+          homepage = "https://github.com/stoeffel/elm-verify-examples";
+          license = licenses.bsd3;
+          maintainers = [ maintainers.turbomack ];
+        };
+      };
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/node/default.nix:22]: this package is currently broken because elmi-to-json is broken due to its dependencies being marked as broken. I think we will probably end up removing elmi-to-json and elm-verify-examples.

File: pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format-test-lib.nix (hunk 1 above)
Comment (PedroHLC) [reply pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format-test-lib.nix:18]: I should be able to rollback, it was mistakenly formatted indeed.

File: pkgs/development/compilers/elm/packages/node/default.nix (hunk 2 above)
Comment (PedroHLC) [reply pkgs/development/compilers/elm/packages/node/default.nix:22]: Let's do this in another PR.

File: pkgs/development/compilers/elm/packages/ghc9_2/elm-format/avh4-lib.nix [hunk 3]
```diff
@@ -22,7 +21,10 @@ mkDerivation {
     relude tasty tasty-hspec tasty-hunit text
   ];
   testToolDepends = [ tasty-discover ];
-  doHaddock = false;
   description = "Common code for haskell projects";
   license = lib.licenses.bsd3;
+
+  # Added manually
+  configureFlags = [ "--ghc-option=-Wno-error=unused-packages" ];
+  doHaddock = false;
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/ghc9_2/elm-format/avh4-lib.nix:29]: I think we can do these in overrides in ghc9_2/default.nix so we don't lose it if the file itself gets regenerated.

Comment (PedroHLC) [reply pkgs/development/compilers/elm/packages/ghc9_2/elm-format/avh4-lib.nix:29]: Sounds much better.

**Comment by PedroHLC** [comment]
Result of `nixpkgs-review pr 299780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>elmPackages.elm-coverage</li>
    <li>elmPackages.elm-verify-examples</li>
    <li>elmPackages.elmi-to-json</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>elmPackages.elm-instrument</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>elmPackages.create-elm-app</li>
    <li>elmPackages.elm</li>
    <li>elmPackages.elm-format</li>
    <li>elmPackages.elm-json</li>
    <li>elmPackages.elm-land</li>
    <li>elmPackages.elm-pages</li>
    <li>elmPackages.elm.doc</li>
  </ul>
</details>

**Comment by PedroHLC** [comment]
Result of `nixpkgs-review pr 299780` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>elmPackages.elm-coverage</li>
    <li>elmPackages.elm-verify-examples</li>
    <li>elmPackages.elmi-to-json</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>elmPackages.create-elm-app</li>
    <li>elmPackages.elm</li>
    <li>elmPackages.elm-format</li>
    <li>elmPackages.elm-instrument</li>
    <li>elmPackages.elm-json</li>
    <li>elmPackages.elm-land</li>
    <li>elmPackages.elm-pages</li>
    <li>elmPackages.elm.doc</li>
  </ul>
</details>

File: pkgs/development/compilers/elm/packages/ghc8_10/elm-instrument/default.nix [hunk 4]
```diff
@@ -37,7 +27,7 @@ mkDerivation {
     quickcheck-io split tasty tasty-golden tasty-hunit tasty-quickcheck
     text
   ];
-  homepage = "https://elm-lang.org";
+  homepage = "http://elm-lang.org";
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/ghc8_10/elm-instrument/default.nix:30]: what is the reason for this change? I think http version will redirect to https anyway.

File: pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format.nix [hunk 5]
```diff
@@ -24,18 +24,8 @@ mkDerivation rec {
...
   mainProgram = "elm-format";
-  postPatch = ''
-    mkdir -p ./generated
-    cat <<EOHS > ./generated/Build_elm_format.hs
-    module Build_elm_format where
-
-    gitDescribe :: String
-    gitDescribe = "${version}"
-    EOHS
-  '';
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format.nix:40]: I think this is probably requored for elm-format to report correct version. Did you test this is really not needed?

cc @avh4

File: pkgs/development/compilers/elm/update.sh [hunk 6]
```diff
@@ -1,13 +1,21 @@
...
-cabal2nix https://github.com/ekmett/ansi-wl-pprint --revision d16e2f6896d76b87b72af7220c2e93ba15c53280 > packages/ansi-wl-pprint.nix
+# Update all cabal packages.
+cabal2nix 'https://github.com/zwilias/elm-instrument' --revision '0.0.7' > packages/ghc8_10/elm-instrument/default.nix
+for subpath in 'avh4-lib' 'elm-format-lib' 'elm-format-markdown' 'elm-format-test-lib'; do
+  cabal2nix 'https://github.com/avh4/elm-format' --revision '0.8.7' \
+    --subpath $subpath > packages/ghc9_2/elm-format/${subpath}.nix
+done
+cabal2nix 'https://github.com/avh4/elm-format' --revision '0.8.7' > packages/ghc9_2/elm-format/elm-format.nix
+cabal2nix 'https://github.com/stoeffel/elmi-to-json' --revision '1.3.0' > packages/ghc8_10/elmi-to-json/default.nix
+cabal2nix 'https://github.com/ekmett/ansi-wl-pprint' --revision 'v0.6.8.1' > packages/ghc9_6/ansi-wl-pprint/default.nix
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/update.sh:12]: good idea to add these :+1:

File: pkgs/development/compilers/elm/packages/ghc8_10/elm-instrument/default.nix (hunk 4 above)
Comment (PedroHLC) [reply pkgs/development/compilers/elm/packages/ghc8_10/elm-instrument/default.nix:30]: Cabal generated it. We're overriding this in a later stage.

File: pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format.nix (hunk 5 above)
Comment (PedroHLC) [reply pkgs/development/compilers/elm/packages/ghc9_2/elm-format/elm-format.nix:40]: We're still applying this, but in the override instead, to keep this file as clean as cargo2nix generated.

**Review by turboMaCk — APPROVED** [review]
LGTM
